[move-only] Fix SIL representation and SILOptimizer to preserve value deinits

atrick opened the pull request:
Fix SIL representation and SILOptimizer to preserve value deinits.

Fixes:
- rdar://110079028 ([move-only] Finish SIL support for drop_deinit)
- rdar://109849028 ([move-only] LowerAggregateInstrs eliminates struct deinitialization)

Many basic SIL passes were silently deleting the struct and enum deinits:
- SROA
- Mem2Reg
- RLE
- DSE
- FunctionSignatureOpts
- LowerAggregates
- SILCombine
- EarlyCodeMotion
- ReleaseHoisting
- *many* passes that rely on ARC analysis (RCIndentity)

Major Changes:
- add SIL verification for drop_deinit
- fix ownership lowering to handle drop_deinit
- disable deinit devirtualization (until the optimizer is tested)
- fix 'shouldExpand' to recognize value deinits; impacts many passes
- fix dead function elimination to collect and preserve value deinits
- fix RCIdentity to recognize value deinits; impacts many passes
- fix SILCombine special cases to preserve value deinits
- fix TypeLowering to preserve value deinits; mainly fixes LowerAggregates

atrick committed:
[move-only] Fix SILOptimizer expansion to preserve deinits.

Many basic SIL passes were silently deleting the struct deinit:
- SROA
- Mem2Reg
- RLE
- DSE
- FunctionSignatureOpts
- LowerAggregates

Fixes rdar://109849028 ([move-only] LowerAggregateInstrs eliminates struct deinitialization)

atrick committed:
[move-only] Fix drop_deinit OSSA lowering

drop_deinit ultimately only affects the semantics of its
destroy_value. Avoid generating releases for destroys in which the
deinit has been dropped. Instead, individually release the members.

atrick committed:
[move-only] Comment emitMoveOnlyMemberDestruction drop_deinit

atrick committed:
[move-only] Verify drop_deinit

drop_deinit only exists in ownership SIL. Remove IRGen support.

A drop_deinit can only ever be destroyed or destructured.

A destructure of a struct-with-deinit requires a drop_deinit operand.

atrick committed:
[move-only] Rename MoveOnlyDeinitInsertion

to MoveOnlyDeinitDevirtualization

atrick committed:
[move-only] Disable move-only devirtualization.

It is not needed for correctness and hides most of the deinit related
optimizer bugs.

gottesmm commented on lib/SIL/Verifier/SILVerifier.cpp:
```diff
@@ -3556,6 +3556,14 @@ class SILVerifier : public SILVerifierBase<SILVerifier> {
                 "destructure with none ownership kind operand and non-none "
                 "ownership kind result?!");
       }
+      if (auto *nominalTy = operandTy.getNominalOrBoundGenericNominal()) {
```
This will always have a nominal ty. I don't think this condition is needed.

gottesmm commented on lib/SILGen/SILGenDestructor.cpp:
```diff
@@ -499,6 +499,8 @@ void SILGenFunction::emitMoveOnlyMemberDestruction(SILValue selfValue,
+  // drop_deinit must be used to invalidate any user-defined struct/enum deinit
+  // before the individual members can be destroyed.
   selfValue = B.createDropDeinit(cleanupLoc, selfValue);
   if (selfValue->getType().isAddress()) {
```
One thing that I don't understand is why here and in the drop_deinit lowering in the ownership model eliminator we are doing destructuring? Seems like it should be in one place, no?

atrick replied:
Ownership lowering needs to correctly lower the destroy_value operation. There's no way around that.

It's always valid to destructure an aggregate that has no deinit. But it shouldn't be done unless needed, so I don't understand why SILGen was doing it. It was only hiding downstream bugs. At this point, I don't think it makes sense to preserve any of the existing behavior, so I went ahead and removed that useless destructure from SILGen.

gottesmm reviewed (commented):
A simplification and a question

gottesmm commented on lib/SILGen/SILGenDestructor.cpp:
```diff
@@ -497,111 +497,59 @@ void SILGenFunction::emitClassMemberDestruction(ManagedValue selfValue,
+      SILValue addr = B.createStructElementAddr(
+          cleanupLoc, selfValue, vd, ti.getLoweredType().getAddressType());
+      addr = B.createBeginAccess(
+          cleanupLoc, addr, SILAccessKind::Deinit, SILAccessEnforcement::Static,
```
Question. Shouldn't we perhaps do the same thing here for destroy_addr like we do for objects. Then we would be really consistent across the code base.

atrick replied:
`drop_deinit` on an address doesn't have the same OSSA semantics as it does on a value. The address and value forms definitely should be separate opcodes to avoid this confusion (in some later PR).

It is analogous to destroy_addr vs. destroy_value.  destroy_addr has nothing to do with OSSA, and drop_deinit on an address does not change the meaning of any subsequent destroy_addr. drop_deinit on an address does not actually do *anything* except block optimization on the address, preventing us from promoting it to OSSA!

So drop_deinit on an address doesn't have any SIL requirements and has no structural verification. As such, SILGen cannot emit a destroy_addr for a discarded value. That would be outright wrong. And OSSA lowering is not affected at all by a drop_deinit on an address.

atrick committed:
[move-only] Fix dead code elimination to preserve value deinits.

atrick committed:
[move-only] Fix SILOptimizer code motion to preserve value deinits

Multiple code motion and ARC related passes were removing struct/enum
deinits.

Passes fixed include:
- SILCombine
- EarlyCodeMotion
- ReleaseHoisting
- *many* passes that rely on ARC analysis (RCIndentity)

atrick committed:
[move-only] Fix SILCombine to preserve value deinit

Fix a special case in visitReleaseValueInst for enum-with-deinit.

atrick committed:
[move-only] Fix TypeLowering to preserve value deinits

MoveOnlyLoadableStruct should only lower to memberwise destroys if it
has no deinit.

atrick committed:
[move-only] Update tests for drop_deinit lowering.

atrick commented:
@swift-ci test

atrick commented:
@swift-ci smoke test windows

atrick committed:
Disable SILGen/discard.swift on Windows.

The -Onone passes fail to remove struct_extract, so the CHECK
lines will fail.

atrick committed:
cleanup whitespace in sil_combine_moveonly.sil

atrick committed:
Cleanup SILCombine hasValueDeinit checks
